### PR TITLE
Introduce plugin signatures

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
This was originally implemented [here](https://github.com/CrowdStrike/ember-headless-table/pull/12), but was extracted due to being a (required) side-quest of that PR in order to support that PR's original goal.

This PR changes the type arguments for all plugins to use a single type argument, called `Signature`, inline with (Modifiers / Helpers / Components / Resources)' nomenclature.

Docs: https://34dd5187.ember-headless-table.pages.dev/api/interfaces/plugins.PluginSignature (hash might change)

### CODE REVIEW

File: ember-headless-table/src/-private/-type-tests/plugins-accessors.test.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  delete(key: string): void | undefined;
+}
+expectTypeOf(preferences.forTable(x, SimplePlugin)).toEqualTypeOf<MapLite>();
+expectTypeOf(preferences.forColumn(y, SimplePlugin)).toEqualTypeOf<MapLite>();
+
+//////////////////////////////////////////////
+// A Real plugin
+//////////////////////////////////////////////
+
+import { DataSorting } from '../../plugins/data-sorting';
```
Comment (danwenzel) [inline ember-headless-table/src/-private/-type-tests/plugins-accessors.test.ts:66]: Curious what allows you to declare `import`s not at the top of the file

File: ember-headless-table/src/-private/-type-tests/plugins-signature-utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    Plugin: AOptions;
+    Column: AColumnOptions;
+  };
+}
+
+expectTypeOf<ColumnOptionsFor<FullFreeformSignature>>().toEqualTypeOf<AColumnOptions>();
+expectTypeOf<OptionsFor<FullFreeformSignature>>().toEqualTypeOf<AOptions>();
+expectTypeOf<TableMetaFor<FullFreeformSignature>>().toEqualTypeOf<ATableMeta>();
+expectTypeOf<ColumnMetaFor<FullFreeformSignature>>().toEqualTypeOf<AColumnMeta>();
+expectTypeOf<RowMetaFor<FullFreeformSignature>>().toEqualTypeOf<ARowMeta>();
```
Comment (danwenzel) [inline ember-headless-table/src/-private/-type-tests/plugins-signature-utils.test.ts:63]: This type testing is quite interesting... Are there any good resources on  the philosophy behind them, best practices, etc?  Or do we wait for you to give a talk on it?  ;)

File: ember-headless-table/src/-private/private-types.ts [hunk 3]
```diff
@@ -2,7 +2,6 @@
  * Private utility types
  */
 
-export interface Constructor<T, Args extends any[] = any[]> {
-  new (...args: Args): T;
-}
+export type Constructor<T, Args extends any[] = any[]> = new (...args: Args) => T;
+// export type Constructor<T, Args extends any[] = any[]> = abstract new (...args: Args) => T;
```
Comment (danwenzel) [inline ember-headless-table/src/-private/private-types.ts:6]: commented line.  Remove? or if not, add comment to explain?

File: ember-headless-table/src/-private/table.ts [hunk 4]
```diff
@@ -178,8 +178,14 @@ export class Table<DataType = unknown> extends Resource<Signature<DataType>> {
   /**
    * Get the active plugin instance for the given plugin class
    */
-  pluginOf(klass: Class<Plugin>) {
-    return this.plugins.find((plugin) => plugin instanceof klass);
+  pluginOf<Instance extends BasePlugin<any>>(klass: Class<Instance>): Instance | undefined {
+    let result = this.plugins.find((plugin) => plugin instanceof klass);
+
+    /**
+     * This is an unsafe cast, because Instance could be unrelated to any of the types
```
Comment (danwenzel) [inline ember-headless-table/src/-private/table.ts:185]: so... if it's unsafe, can we explain why we are doing it?

**Comment by danwenzel** [comment]
Really nice improvements here! 

So what will the migration path look like for someone upgrading to this version?  I'm assuming this'll be a major bump, right?

**Review by danwenzel — APPROVED** [review]

File: ember-headless-table/src/plugins/column-resizing/plugin.ts [hunk 5]
```diff
@@ -48,14 +48,22 @@ export interface TableOptions {
   handlePosition?: string;
 }
 
+interface Signature {
```
Comment (nicolechung) [inline ember-headless-table/src/plugins/column-resizing/plugin.ts:51]: Much nicer to read!

**Review by ynotdraw — APPROVED** [review]
🚀

File: ember-headless-table/src/-private/-type-tests/plugins-accessors.test.ts (hunk 1 above)
Comment (NullVoxPopuli) [reply ember-headless-table/src/-private/-type-tests/plugins-accessors.test.ts:66]: the import linting plugin (same as we use internally) starts a new set of groups after a comment

File: ember-headless-table/src/-private/-type-tests/plugins-signature-utils.test.ts (hunk 2 above)
Comment (NullVoxPopuli) [reply ember-headless-table/src/-private/-type-tests/plugins-signature-utils.test.ts:63]: Not sure, tbh. I learned everything from
 - https://github.com/mmkal/expect-type
 - the ember discord
 - and getting tired of manually testing type inference / intellisense 😅

File: ember-headless-table/src/-private/private-types.ts (hunk 3 above)
Comment (NullVoxPopuli) [reply ember-headless-table/src/-private/private-types.ts:6]: oh! yup, this can be removed

**Review by nicolechung — APPROVED** [review]

File: ember-headless-table/src/plugins/column-resizing/plugin.ts (hunk 5 above)
Comment (NullVoxPopuli) [reply ember-headless-table/src/plugins/column-resizing/plugin.ts:51]: thanks!

File: ember-headless-table/src/plugins/column-visibility/plugin.ts [hunk 6]
```diff
@@ -21,40 +21,44 @@ declare module 'ember-headless-table/plugins' {
...
+    };
+    Column: {
+      /**
+       * The default visibilty of the column, when rendered.
+       * The column can still be toggled on and off.
+       *
+       * When interacting with preferences, the value stored in preferenced
+       * will be the inverse of this value (to save space in storage).
+       */
+      isVisible?: boolean;
```
[the remark this answers is not in the document]
Comment (NullVoxPopuli) [reply ember-headless-table/src/plugins/column-visibility/plugin.ts:41]: thanks! 🎉

**Comment by NullVoxPopuli** [comment]
> So what will the migration path look like for someone upgrading to this version? I'm assuming this'll be a major bump, right?

Naw, internally what I've been doing is:
 - release initial version in OSS (1.0.0)
 - setup PR that uses the OSS version  (we use `*` internally to mean "latest", so our lockfile ends up pointing at the correct registry with this type of change)
 - merge
 - delete package in internal registry

**Pull request merged**